File: 2015/day7.py
import AOCInit
import util

if __name__ != '__main__':
    exit()

inp = util.getInput(d=7, y=2015).splitlines()

opDict = {
    'AND': (lambda x, y: x & y),
    'OR': (lambda x, y: x | y),
    'LSHIFT': (lambda x, y: (x << y) % 65536),
    'RSHIFT': (lambda x, y: x >> y),
    'ID': (lambda x: x),
    'NOT': (lambda x: 65536 + ~x),
}
valDict = dict()
depDict = dict()

# part 1
for inst in inp:
    (cmd, target) = inst.split(' -> ', maxsplit=1)
    cmdList = cmd.split(' ')
    if len(cmdList) == 1:
        # assignment
        depDict[target] = ('ID', (cmdList[0], ))
    elif len(cmdList) == 2:
        # not
        depDict[target] = ('NOT', (cmdList[1], ))
    else:
        # other
        depDict[target] = (cmdList[1], (cmdList[0], cmdList[2]))

def getVal(val):
    if val[0].isdigit():
        return int(val)
    else:
        if val not in valDict:
            dep = depDict[val]
            valDict[val] = opDict[dep[0]](*map(getVal, dep[1]))
        return valDict[val]

aVal = getVal('a')
print(aVal)

# part 2
valDict = {'b': aVal}
print(getVal('a'))

File: 2015/day7.rb
class Circut
  @@Symb_map = {
    'AND' => :&,
    'OR' => :|,
    'RSHIFT' => :>>,
    'LSHIFT' => :<<
  }.freeze

  def initialize
    @dependency = Hash.new
    @eval_res = Hash.new
  end

  def add_rec(line)
    x, y = line.split(' -> ')
    @dependency[y] = x.split
  end

  def cal(name)
    if name.match? /\d+/
      return name.to_i
    end

    unless @eval_res.include? name
      dep = @dependency[name]
      @eval_res[name] = \
        case dep.size
        when 1
          dep[0].match?(/\d+/) ? dep[0].to_i : cal(dep[0])
        when 2
          65536 + ~cal(dep[1])
        when 3
          cal(dep[0]).send(@@Symb_map.fetch(dep[1]), cal(dep[2]))
        end
    end
    @eval_res.fetch name
  end

  def reset
    @eval_res.clear
  end
end

inp = File.read('input7')

# part 1
circ = Circut.new
inp.each_line { |line| circ.add_rec line.chomp }
a_val = circ.cal 'a'
puts a_val

# part 2
circ.reset
circ.add_rec "#{a_val} -> b"
puts circ.cal 'a'
